src/coreclr/src/jit/importer.cpp

```diff
@@ -5941,6 +5941,34 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken, const B
                                 }
                             }
                         }
+                        break;
+
+                    // box + isinst + unbox.any
+                    case CEE_UNBOX_ANY:
+                        if ((nextCodeAddr + 1 + sizeof(mdToken)) <= codeEndp)
+                        {
+                            CORINFO_RESOLVED_TOKEN isinstResolvedToken = {};
+                            impResolveToken(codeAddr + 1, &isinstResolvedToken, CORINFO_TOKENKIND_Class);
+
+                            CORINFO_RESOLVED_TOKEN unboxResolvedToken = {};
```

Comment:
The callbacks from the jit to the runtime have some cost.  Can you first check if isinst and box agree and only if that is so, resolve the unbox token and check it against the box.